javascript/webdriver/touchsequence.js

```diff
@@ -0,0 +1,280 @@
+// Copyright 2015 Selenium comitters
+// Copyright 2015 Software Freedom Conservancy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+goog.provide('webdriver.TouchSequence');
+
+goog.require('goog.array');
+goog.require('webdriver.Command');
+goog.require('webdriver.CommandName');
+
+
+
+/**
+ * Class for defining sequences of user touch interactions. Each sequence
+ * will not be executed until {@link #perform} is called.
+ *
+ * <p>Example:<pre><code>
+ *   new webdriver.TouchSequence(driver).
+ *       tapAndHold({x: x1, y: y1}).
+ *       move({x: x2, y: y2}).
+ *       release({x: x3, y: y3}).
+ *       perform();
+ * </pre></code>
+ *
+ * @param {!webdriver.WebDriver} driver The driver instance to use.
+ * @constructor
+ */
+webdriver.TouchSequence = function(driver) {
+
+  /** @private {!webdriver.WebDriver} */
+  this.driver_ = driver;
+
+  /** @private {!Array.<{description: string, command: !webdriver.Command}>} */
+  this.touchActions_ = [];
+};
+
+
+/**
+ * Schedules an action to be executed each time {@link #perform} is called on
+ * this instance.
+ * @param {string} description A description of the command.
+ * @param {!webdriver.Command} command The command.
+ * @private
+ */
+webdriver.TouchSequence.prototype.schedule_ = function(description, command) {
+  this.touchActions_.push({
+    description: description,
+    command: command
+  });
+};
+
+
+/**
+ * Executes this action sequence.
+ * @return {!webdriver.promise.Promise} A promise that will be resolved once
+ *     this sequence has completed.
+ */
+webdriver.TouchSequence.prototype.perform = function() {
+  // Make a protected copy of the scheduled actions. This will protect against
+  // users defining additional commands before this sequence is actually
+  // executed.
+  var actions = goog.array.clone(this.touchActions_);
+  var driver = this.driver_;
+  return driver.controlFlow().execute(function() {
+    goog.array.forEach(actions, function(action) {
+      driver.schedule(action.command, action.description);
+    });
+  }, 'TouchSequence.perform');
+};
+
+
+/**
+ * Taps an element.
+ *
+ * @param {webdriver.WebElement} elem The element to tap.
```

Comment:
Type annotation must be `{!webdriver.WebElement}` since WebElement is non-nullable. Please fix throughout.